pyquil/api/_qpu.py

```diff
@@ -71,34 +73,68 @@ def _extract_bitstrings(ro_sources: List[Optional[Tuple[int, int]]],
 
 class QPU(QAM):
     @_record_call
-    def __init__(self, endpoint: str, user: str = "pyquil-user", priority: int = 1) -> None:
+    def __init__(self,
+                 endpoint: Optional[str] = None,
+                 user: str = "pyquil-user",
+                 priority: int = 1,
+                 config: Optional[PyquilConfig] = None) -> None:
         """
         A connection to the QPU.
 
-        :param endpoint: Address to connect to the QPU server.
-        :param user: A string identifying who's running jobs.
-        :param priority: The priority with which to insert jobs into the QPU queue. Lower
+        :param endpoint: Address to connect to the QPU server. If not provided, the
+            endpoint provided by engagement with dispatch is used. One or both must be
+            available and valid.
+        :param user: [deprecated] A string identifying who's running jobs.
+        :param priority: [deprecated] The priority with which to insert jobs into the QPU queue. Lower
                          integers correspond to higher priority.
+        :param config: PyQuilConfig object, which provides endpoint & engagement values.
         """
+        if config:
+            self.config = config
+        else:
+            self.config = PyquilConfig()
+
+        self.endpoint = endpoint
+        self._last_results: Dict[str, np.ndarray] = {}
+
+        super().__init__()
 
+    def build_client(self) -> Client:
+        endpoint = self.endpoint or self.config.qpu_url
         if endpoint is None:
-            raise RuntimeError("""It looks like you've tried to run a program against a QPU but do
- not currently have a reservation on one. To reserve time on Rigetti
- QPUs, use the command line interface, qcs, which comes pre-installed
- in your QMI. From within your QMI, type:
+            raise UserMessageError(
+                """It looks like you've tried to run a program against a QPU but do
+                not currently have a reservation on one. To reserve time on Rigetti
```

Comment:
If this is actually going to print out to a user

```
It looks like you've tried to run a program against a QPU but do
                not currently have a reservation on one. To reserve time on Rigetti
                ...
```
then we should strip the extra indentation